dimod/generators/satisfiability.py

```diff
@@ -0,0 +1,137 @@
+
+import collections.abc
+import itertools
+import math
+import typing
+
+import numpy as np
+
+import dimod  # for typing
+
+from dimod.binary import BinaryQuadraticModel
+from dimod.vartypes import Vartype
+
+
+__all__ = ["nae3sat"]
+
+
+def _iter_interactions(num_variables: int, num_clauses: int,
+                       seed: typing.Union[None, int, np.random.Generator] = None,
+                       ) -> typing.Iterator[typing.Tuple[int, int, int]]:
+
+    rng = np.random.default_rng(seed)
+
+    for _ in range(num_clauses):
+        x, y, z = rng.choice(num_variables, 3, replace=False)
+        xs, ys, zs = 2 * rng.integers(0, 1, endpoint=True, size=3) - 1
+        yield (x, y, xs*ys)
+        yield (x, z, xs*zs)
+        yield (y, z, ys*zs)
+
+
+def _iter_interactions_without_replacement(
+        num_variables: int, num_clauses: int,
+        seed: typing.Union[None, int, np.random.Generator] = None,
+        ) -> typing.Iterator[typing.Tuple[int, int, int]]:
+
+    try:
+        max_num_clauses = math.comb(num_variables, 3)
+    except AttributeError:
+        # Python < 3.8
+        # This is a bit less picky and has overflow issues but
+        # probably fine for any cases users are likely to actually
+        # submit
+        max_num_clauses = math.factorial(num_variables)
+        max_num_clauses //= (6 * math.factorial(num_variables - 3))
+
+    if num_clauses > max_num_clauses:
+        raise ValueError("rho results in too many clauses")
+
+    rng = np.random.default_rng(seed)
+
+    seen: typing.Set[typing.FrozenSet[int]] = set()
+    while len(seen) < num_clauses:
+        x, y, z = rng.choice(num_variables, 3, replace=False)
+
+        fz = frozenset((x, y, z))  # so we're order-invarient
+        if fz in seen:
+            continue
+        seen.add(fz)
+
+        xs, ys, zs = 2 * rng.integers(0, 1, endpoint=True, size=3) - 1
+
+        yield (x, y, xs*ys)
+        yield (x, z, xs*zs)
+        yield (y, z, ys*zs)
+
+
+def nae3sat(variables: typing.Union[int, typing.Sequence[dimod.typing.Variable]],
+            rho: float = 2.1,
+            *,
+            seed: typing.Union[None, int, np.random.Generator] = None,
+            replace: bool = True,
+            ) -> BinaryQuadraticModel:
+    """Generator for Not-All-Equal 3-SAT (NAE3SAT) Binary Quadratic Models.
+
+    NAE3SAT_ is an NP-complete problem class that consists in satistying a number of conjunctive
+    clauses that involve three variables (or variable negations). The variables on each clause
+    should be not-all-equal. Ie. all solutions except ``(+1, +1, +1)`` or
+    ``(-1, -1, -1)`` are valid for each class.
+
+    .. _NAE3SAT: https://en.wikipedia.org/wiki/Not-all-equal_3-satisfiability
+
+    Args:
+        num_variables: The number of variables in the problem.
+        rho: The clause-to-variable ratio.
+        seed: Passed to :func:`numpy.random.default_rng()`, which is used
+            to generate the clauses and the negations.
```

Comment:
why the "and the negations" addition to "clauses"? the statement doesn't talk about variables